raft/tracker/progress.go

```diff
@@ -59,6 +59,12 @@ type Progress struct {
 	// true, raft should pause sending replication message to this peer until
 	// ProbeSent is reset. See ProbeAcked() and IsPaused().
 	ProbeSent bool
+	// MsgAppSent is used while this follower is in StateReplicate but the flow of
+	// messages is throttled, i.e. Inflights is full. In this state we still need
+	// to send empty MsgApp to guarantee progress, but we do so only when
+	// MsgAppSent == false.
+	// This works similarly to ProbeSent, only in a different state.
+	MsgAppSent bool
```

Comment:
We could probably use the same `MsgAppSent` here both for probe and replicate state. The technique for both cases is the same, and boils down to "probes" on heartbeats.

Reply:
Added a commit after this one implementing this idea. Please let me know which way it looks better. I prefer re-using the same bool, cause it's the same heuristic.

Reply:
+1 on keeping only one field, but I found `MsgAppSent` confusing. Raft sends MsgApp all the time; if this field is specific to throttling let's give it a more fitting name. I think ProbeSent actually made more sense: in the steady state, are not necessary because the follower is acking writes before inflight buffer fills up. We send probes periodically; the field is reset on each heartbeat or when a probe is successful (which means we don't need to probe any more). Probes can occur in two states: either in StateProbe, where we are trying to find out at what index the follower can accept appends (and here I think we're sending full log entries), or when inflights is full (in which case we send an empty append, just to make sure the follower can give us an ack or rej back in case previous msgs got dropped).

Reply:
In particular, `IsPaused()` just becomes a read on `ProbeSent` and we don't need the awkward conjunction with `Inflights.Full()`. Instead, when we see that inflights are full, and we send a probe, we set ProbeSent if it isn't already set. That seems simpler because now the field directly controls what is going on.

Reply:
Removed the rename commit, it's `ProbeSent` again. Will try your proposal tomorrow.

Reply:
Trying this proposal. The thing is, when we remove the awkward conjunction in one place, it reappears in other places (we should check both `Full()` and "recently sent" facts anyway), so the net awkwardness reduction is 0. I think we should just push a bit more stuff to `Progress` type, as noted in a TODO.

Reply:
Added another commit with a better encapsulated logic in `Progress`.